libraries/chain/block_state.cpp

```diff
@@ -102,5 +102,34 @@ namespace eosio::chain {
    ,_cached_trxs( std::move(trx_metas) )
    {}
 #endif
+
+   std::optional<qc_data_t> block_state::get_best_qc() const {
+      auto block_number = block_num();
+
+      // if pending_qc does not have a valid QC, consider valid_qc only
+      if( !pending_qc.is_quorum_met() ) {
+         if( valid_qc ) {
+            return qc_data_t{ quorum_certificate{ block_number, *valid_qc },
+                              qc_info_t{ block_number, valid_qc->is_strong() }};
+         } else {
+            return std::nullopt;;
+         }
+      }
+
+      // extract valid QC from pending_qc
+      valid_quorum_certificate valid_qc_from_pending(pending_qc);
+
+      // if valid_qc does not have value, consider valid_qc_from_pending only
+      if( !valid_qc ) {
+         return qc_data_t{ quorum_certificate{ block_number, valid_qc_from_pending },
+                           qc_info_t{ block_number, valid_qc_from_pending.is_strong() }};
+      }
+
+      // Both valid_qc and valid_qc_from_pending have value. Compare them and select a better one.
+      // Strong beats weak. Break tie with highest accumulated weight.
```

Comment:
Don't yet break the tie with the highest accumulated weight. Just pick `valid_qc` if both are strong.